controllers/cruisecontroltask_controller.go

```diff
@@ -62,448 +59,412 @@ func (r *CruiseControlTaskReconciler) Reconcile(ctx context.Context, request ctr
-				parsedTime, err := ccutils.ParseTimeStampToUnixTime(volumeState.TaskStarted)
-				if err != nil {
-					return errors.WrapIf(err, "could not parse timestamp")
+	for brokerId, brokerStatus := range instance.Status.BrokersState {
+		if brokerStatus.GracefulActionState.CruiseControlState.IsActionRequired() {
+			state := brokerStatus.GracefulActionState
+			switch {
+			case state.CruiseControlState.IsUpscale():
+				t := &CruiseControlTask{
+					TaskID:      state.CruiseControlTaskId,
+					BrokerID:    brokerId,
+					BrokerState: state.CruiseControlState,
+					StartedAt:   state.TaskStarted,
+					Operation:   OperationAddBroker,
+					Err:         state.ErrorMessage,
 				}
-
-				if time.Since(parsedTime).Minutes() > kafkaCluster.Spec.CruiseControlConfig.CruiseControlTaskSpec.GetDurationMinutes() {
-					volumesStateWithTimedOutDiskCCTask[mountPath] = kafkav1beta1.VolumeState{
-						CruiseControlVolumeState: kafkav1beta1.GracefulDiskRebalanceRequired,
-						CruiseControlTaskId:      volumeState.CruiseControlTaskId,
-						ErrorMessage:             "Timed out waiting for the disk rebalance cc task to complete",
-						TaskStarted:              volumeState.TaskStarted,
-					}
+				tasksAndStates.Add(t)
+			case state.CruiseControlState.IsDownscale():
+				t := &CruiseControlTask{
+					TaskID:      state.CruiseControlTaskId,
+					BrokerID:    brokerId,
+					BrokerState: state.CruiseControlState,
+					StartedAt:   state.TaskStarted,
+					Operation:   OperationRemoveBroker,
+					Err:         state.ErrorMessage,
 				}
+				tasksAndStates.Add(t)
 			}
 		}
-		if len(volumesStateWithTimedOutDiskCCTask) > 0 {
-			brokersWithTimedOutCCTask = append(brokersWithTimedOutCCTask, brokerId)
-			brokersVolumesStateWithTimedOutDiskCCTask[brokerId] = volumesStateWithTimedOutDiskCCTask
-		}
-	}
 
-	// task timed out
-	if len(brokersWithTimedOutCCTask) > 0 {
-		log.Info("Killing Cruise control task", "taskId", ccTaskId)
-		cc := scale.NewCruiseControlScaler(kafkaCluster.Namespace, kafkaCluster.Spec.GetKubernetesClusterDomain(), kafkaCluster.Spec.CruiseControlConfig.CruiseControlEndpoint, kafkaCluster.Name)
-
-		err = cc.KillCCTask()
-		if err != nil {
-			return errorfactory.New(errorfactory.CruiseControlNotReady{}, err, "cc communication error")
-		}
-		err = k8sutil.UpdateBrokerStatus(r.Client, brokersWithTimedOutCCTask, kafkaCluster, brokersVolumesStateWithTimedOutDiskCCTask, log)
-		if err != nil {
-			return errors.WrapIfWithDetails(err, "could not update status for broker(s)", "id(s)", strings.Join(brokersWithTimedOutCCTask, ","))
+		for mountPath, volumeState := range brokerStatus.GracefulActionState.VolumeStates {
+			if volumeState.CruiseControlVolumeState.IsActionRequired() {
+				t := &CruiseControlTask{
+					TaskID:      volumeState.CruiseControlTaskId,
+					BrokerID:    brokerId,
+					StartedAt:   volumeState.TaskStarted,
+					Volume:      mountPath,
```

Comment:
I wonder if `Volume` should be renamed to `KafkaLogDir` for example.

Reply:
I agree that `KafkaLogDir` would be more appropriate however I would stick to the `Volume` as the state for it is described as `VolumeState`. The renaming would just make it more confusing.